lib/editor/components/ExceptionDateRange.js

```diff
@@ -0,0 +1,162 @@
+import {DropdownButton, MenuItem} from 'react-bootstrap'
+import DateTimeField from 'react-bootstrap-datetimepicker'
+import moment from 'moment'
+
+import {updateActiveGtfsEntity} from '../actions/active'
+import {modifyRangeOfDates, updateDates} from '../../common/util/exceptions'
+import type {ScheduleException, ScheduleExceptionDateRange} from '../../types'
+
+import {inputStyleProps} from './ExceptionDate'
+
+type Props = {
+  activeComponent: string,
+  activeEntity: ScheduleException,
+  index: number,
+  isInvalid: boolean,
+  range: ScheduleExceptionDateRange,
+  updateActiveGtfsEntity: typeof updateActiveGtfsEntity
+}
+
+export default class ExceptionDateRange extends Component<Props> {
+  // This is kept as an arrow function in order to access the component's `this`
+  _onRangeUpdate = (millis: number, isStartDate: boolean) => {
+    const {activeComponent, activeEntity, range, updateActiveGtfsEntity} = this.props
+    const {endDate, startDate} = range
+
+    const newRangeBoundary = moment(+millis)
+    const startMoment = moment(startDate)
+    const endMoment = moment(endDate)
+
+    // This prevents the user from inputting invalid data. Validation is not required since this behaviour
+    // is fully prevented (see also use of minDate and maxDate in DateTimePickers)
+    if (
+      (isStartDate && (startDate === newRangeBoundary.format('YYYYMMDD') || (isStartDate && newRangeBoundary.diff(endMoment) > 0))) ||
+      (!isStartDate && (endDate === newRangeBoundary.format('YYYYMMDD') || newRangeBoundary.diff(startMoment) < 0))
+    ) {
+      return
+    }
+
+    // Filter out new dates (0s), add them back in later to preserve new dates in final array
+    const newDates = activeEntity.dates.filter(date => typeof date === 'number')
+    let dates = activeEntity.dates.filter(date => typeof date === 'string')
+
+    // Check the updated bounds are larger than before then add or remove dates accordingly.
+    // $FlowFixMe: Flow doesn't recognise array filter above.
+    dates = updateDates(isStartDate, newRangeBoundary, dates, startMoment, endMoment)
+
+    updateActiveGtfsEntity({
+      component: activeComponent,
+      entity: activeEntity,
+      props: {dates: [...dates, ...newDates]}
+    })
+  }
+
+  _onSelect = (key: string) => {
+    const {activeComponent, activeEntity, range, updateActiveGtfsEntity} = this.props
+    let dates = [...activeEntity.dates]
+
+    const newDates = activeEntity.dates.filter(date => typeof date === 'number')
+    // $FlowFixMe Flow does not recognize filtering on typeof
+    dates = (dates.filter(date => typeof date === 'string'): Array<string>)
+
+    const startMoment = moment(range.startDate)
+    const endMoment = moment(range.endDate)
+
+    if (key === 'range') {
+      dates = modifyRangeOfDates(dates, endMoment, startMoment, true)
+    } else if (key === 'end-date') {
+      dates = modifyRangeOfDates(dates, endMoment, startMoment.add(1, 'days'), true)
+    }
```

Comment:
This might be cleaner as one function call that uses variables that are set based on the key

Reply:
good call, this is much cleaner in https://github.com/ibi-group/datatools-ui/pull/828/commits/b8ea974dc8ba296b92d2b964760b893be231af36